presto-main/src/main/java/com/facebook/presto/operator/PagesRTreeIndex.java

```diff
@@ -0,0 +1,193 @@
+    private static final int ABSTRACT_NODE_INSTANCE_SIZE = ClassLayout.parseClass(AbstractNode.class).instanceSize();
+    private static final long[] EMPTY_ADDRESSES = new long[0];
+
+    private final List<Type> types;
+    private final List<Integer> outputChannels;
+    private final List<List<Block>> channels;
+    private final STRtree rtree;
+    private final BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest;
+    private final JoinFilterFunction filterFunction;
+    private final long memorySizeInBytes;
+
+    public static final class GeometryWithAddress
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(GeometryWithAddress.class).instanceSize();
+        private final OGCGeometry ogcGeometry;
+        private final long address;
+
+        public GeometryWithAddress(OGCGeometry ogcGeometry, long address)
+        {
+            this.ogcGeometry = ogcGeometry;
+            this.address = address;
+        }
+
+        public OGCGeometry getOgcGeometry()
+        {
+            return ogcGeometry;
+        }
+
+        public long getAddress()
+        {
+            return address;
+        }
+
+        public long getEstimatedMemorySizeInBytes()
+        {
+            return INSTANCE_SIZE + GeometryUtils.getEstimatedMemorySizeInBytes(ogcGeometry);
+        }
+    }
+
+    public PagesRTreeIndex(
+            Session session,
+            LongArrayList addresses,
+            List<Type> types,
+            List<Integer> outputChannels,
+            List<List<Block>> channels,
+            STRtree rtree,
+            BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest,
+            Optional<JoinFilterFunctionFactory> filterFunctionFactory)
+    {
+        this.types = types;
+        this.outputChannels = outputChannels;
+        this.channels = requireNonNull(channels, "channels is null");
+        this.rtree = requireNonNull(rtree, "rtree is null");
+        this.spatialRelationshipTest = requireNonNull(spatialRelationshipTest, "spatial relationship is null");
+        this.filterFunction = filterFunctionFactory.map(factory -> factory.create(session.toConnectorSession(), addresses, channels)).orElse(null);
+        this.memorySizeInBytes = computeMemorySizeInBytes(addresses);
+    }
+
+    private static Envelope getEnvelope(OGCGeometry ogcGeometry)
+    {
+        com.esri.core.geometry.Envelope env = new com.esri.core.geometry.Envelope();
+        ogcGeometry.getEsriGeometry().queryEnvelope(env);
+
+        return new Envelope(env.getXMin(), env.getXMax(), env.getYMin(), env.getYMax());
+    }
+
+    private long computeMemorySizeInBytes(LongArrayList addresses)
+    {
+        return INSTANCE_SIZE + STRTREE_INSTANCE_SIZE + computeMemorySizeInBytes(rtree.getRoot()) +
+                sizeOf(addresses.elements()) + channels.stream()
+                    .flatMap(List::stream)
+                    .mapToLong(Block::getRetainedSizeInBytes)
```

Comment:
`channels` is a list of lists of blocks. You calculate blocks size, but i think you should also calculate top-level list size and the contained lists sizes as well. (In the extreme theoretical case you can have very long list of empty lists)

Reply:
@findepi @dain This calculation matches `SimplePagesHashStrategy#getSizeInBytes`. Should I update both places?

Reply:
i'd defer to @dain